controller/src/main/java/io/pravega/controller/server/bucket/BucketManagerLeader.java

```diff
@@ -0,0 +1,191 @@
+    @Getter
+    private final BucketDistributor bucketDistributor;
+    //Service type
+    private final BucketStore.ServiceType serviceType;
+
+    public BucketManagerLeader(BucketStore bucketStore, int minRebalanceInterval, BucketDistributor bucketDistributor,
+                               BucketStore.ServiceType serviceType) {
+        Preconditions.checkNotNull(bucketStore, "bucketStore");
+        Preconditions.checkArgument(minRebalanceInterval >= 0, "minRebalanceInterval should not be negative");
+
+        this.bucketStore = bucketStore;
+        this.minRebalanceInterval = Duration.ofSeconds(minRebalanceInterval);
+        this.bucketDistributor = bucketDistributor;
+        this.serviceType = serviceType;
+    }
+
+    /**
+     * Suspend the leader thread.
+     */
+    public void suspend() {
+        suspended.set(true);
+    }
+
+    /**
+     * Resume the suspended leader thread.
+     */
+    public void resume() {
+        if (suspended.compareAndSet(true, false)) {
+            suspendMonitor.release();
+        }
+    }
+
+    @Override
+    public void takeLeadership(CuratorFramework client) throws Exception {
+        log.info("{}: Obtained leadership to monitor the controller to buckets Mapping", serviceType);
+
+        //Attempt a distribution whenever leadership is obtained to ensure no Controllers events are missed.
+        controllerChange.release();
+
+        //Start cluster monitor.
+        pravegaServiceCluster = new ClusterZKImpl(client, ClusterType.CONTROLLER);
+
+        //Add listener to track controller changes on the monitored pravega cluster.
+        pravegaServiceCluster.addListener((type, controller) -> {
+            switch (type) {
+                case HOST_ADDED:
+                case HOST_REMOVED:
+                    //We don't keep track of the controllers and we always query for the entire set from the cluster
+                    //when changes occur. This is to avoid any inconsistencies if we miss any notifications.
+                    log.info("{}: Received event: {} for host: {}. Wake up leader for distribution", serviceType, type, controller);
+                    controllerChange.release();
+                    break;
+                case ERROR:
+                    //This event should be due to ZK connection errors and would have been received by the monitor too,
+                    //hence not handling it explicitly here.
+                    log.info("{}: Received error event when monitoring the pravega host cluster, ignoring...", serviceType);
+                    break;
+            }
+        });
+
+        //Keep looping here as long as possible to stay as the leader and exclusively monitor the pravega cluster.
+        while (true) {
+            try {
+                if (suspended.get()) {
+                    log.info("{}: Monitor is suspended, waiting for notification to resume", serviceType);
+                    suspendMonitor.acquire();
```

Comment:
If due to any reason, zookeeper connection get suspended then this semaphore will keep current thread in suspended state.

Reply:
No, it can not happen because suspend and resume method is dependent on zookeeper notification. We have registered a listener on  zookeeper state and on the basis of state change we are explicitly calling these method in `ZooKeeperBucketManager ` in `startLeaderElection()`. If due to any reason, zookeeper connection get suspended, then it will invoke the suspend() in leaderElectionListener and same for resume().
So there is no possibility.

Reply:
> The only place where suspendMonitor is released is in resume() method, which seems not to be covered by tests

Improved test case.